### PR TITLE
Kb layout switch

### PR DESCRIPTION
**Pull request by matthieuriolo**
As promised to @le-jzr in PR [168](https://github.com/HelenOS/helenos/pull/168) here is a new PR which keeps the hardcoded layouts but provides the needed IPC and command for switching between the layouts. LCTRL-F1 still allows you to switch to the US layout. Unlike to the previous PR the default keyboard can not any longer be configured with `make config`. The idea is that in the future (once keymaps are implemented) the default layout/map will still be hardcoded to ensure that there is at least one keymap around.

The following wikipages need to be updated:
http://www.helenos.org/wiki/UsersGuide/Console
http://www.helenos.org/wiki/Tutorial

Sidenote, both wiki pages are missing the layout "ar" for arabic and it only works if the left CTRL is pressed

-----------------------------

For the keymap, I looked around and found out that it appears every operating system seems to be doing their own thing. I looked at FreeBSD, Arch Linux, Mac OS X and Windows 10. From what I can tell, the linux keymap is the most powerful while FreeBSD has the most simplest one. Linux provides the functionality:
 - Including/Loading other keymaps
 - Single character mapping
 - Auto uppercasing (ex.: +a will automatically map to A when shift/caps is pressed)
 - Mapping to a string instead of a single character
 - Strings can contain a newline and will therefore execute in a shell
 - Dead keys
 - able to differentiate between left/right ctrl or shift

Looking at those functionality I dont believe the structure of HelenOS is able to cope with them. And also probably shouldnt provide those functionality. As result of that I would have suggested to implement a slighty changed version of the FreeBSD keymap

```
# scancode base shift ctrl ctrl+shift alt alt+shift alt+ctrl alt+ctrl+shift   lock state
0x1e 'a' 'A' 1 1 'a' 'A' 1 1 caps
96 gravel gravel - - - - - - -
gravel          32   '`'      'a'  'à'      0xE8   'è'      'i'  0xEC
```
https://github.com/freebsd/freebsd/blob/master/share/vt/keymaps/fr.acc.kbd

 - control characters have to be expressed by their numeric value
 - lock state can be: -, caps, num, both (instead of C, N, B, O)
 - custom "accent names" (must start with a letter; single world)
 - accent mapping is done by a key value array (instead of using brackets)
 - the minus symbol can be used to dismiss an option (instead of nop)

### CODE REVIEW

File: uspace/srv/hid/input/input.c
Comment (jxsvoboda) [inline uspace/srv/hid/input/input.c]: You probably forgot to remove this.

File: uspace/app/layout/layout.c
Comment (jxsvoboda) [inline uspace/app/layout/layout.c]: I think rather the TODO should say we should query the server for available maps.

File: uspace/srv/hid/input/input.c
Comment (jxsvoboda) [inline uspace/srv/hid/input/input.c]: Please don't take away the ability to switch between layouts via keyboard shortcuts.

**Comment by matthieuriolo** [comment]
> while it's nice to have the ability to query and set the layout via the command line in addition to keyboard shortcut, it's not a replacement, so that functionality should still be there, please don't remove it. We could do a shortcut for cycling the layouts.

I would strongly advice against having the functionality of "cycling". This would not work well in the future when the keymaps are implemented as files. I doubt that anyone really needs this feature? The old implementation has the obvious problem of a limited support of 9 keymaps. I think it would be more important to free up those CTRL-combinations for more important functionality. Ex.: Shutdown, reboot, new BDSH Terminal, etc.

> Even then we could keep the Fns for some select keymaps (which could be made configurable in the future)

The prober way to approach this would be to implement shortcut-functionality. Selecting a specific keymap would be a shortcut. However this is heavily depended on what type of keymap we use

They way I see it, this is legacy code which should be removed (better sooner as later)

File: uspace/app/layout/layout.c
Comment (matthieuriolo) [reply uspace/app/layout/layout.c]: When I implemented this function I thought about this and came to the conclusion that querying the server is unnecessary complicated with no benefit. It's an extra IPC call with no real advantage? I think it's easier if the server receives a path with the location of the keymap. Displaying a list (and translating that list) is a job done by the app

**Comment by jxsvoboda** [comment]
> I would strongly advice against having the functionality of "cycling". This would not work well in the future when the keymaps are implemented as files. I doubt that anyone really needs this feature?
What I meant is cycling between a user-preselected set of keymaps, not between all available keymaps. This feature is present just about in every desktop environment and most people who use multiple keymaps use it.

> The old implementation has the obvious problem of a limited support of 9 keymaps. I think it would be more important to free up those CTRL-combinations for more important functionality. Ex.: Shutdown, reboot, new BDSH Terminal, etc.
Well we don't have anywhere near 9 keymaps right now.

> The way I see it, this is legacy code which should be removed (better sooner as later)
Code that provides useful functionality should not be removed without providing appropriate replacement.

**Comment by matthieuriolo** [comment]
I believe this merge request has been forgotten. Keep in mind that french keyboard has been added in the meanwhile.

**Comment by le-jzr** [comment]
Sorry about that, I've been super busy in the past forever and completely forgot about this. Also, there seem to be conflicts now, which need to be resolved. One is caused by the switch to Meson build system, should be an easy fix but let me know if you need hints. Other than that, @jxsvoboda is currently working on the general UI area, so I defer to him.

**Comment by matthieuriolo** [comment]
> Sorry about that, I've been super busy in the past forever and completely forgot about this

Nothing to be sorry about. I do know and understand that you guys are powerhorsing an operating system in your free time and therefore such small changes get forgotten 😃

> Also, there seem to be conflicts now, which need to be resolved. One is caused by the switch to Meson build system, should be an easy fix but let me know if you need hints

Just that we are on the same page: You want me to merge master, resolve the issues and push force the new pull request? I remember that jakub once told me that I should prevent unnecessary merge commits and that it's better to let you guys handle such small conflicts. Just asking

**Comment by jermar** [comment]
I am not sure I said exactly that :-) IMHO the best approach would be rebasing your change on top of the current master and force-pushing it into your branch. This way there won't be the extra commit which merges master into your branch. So in total two merge commits will be avoided.

File: uspace/app/layout/layout.c
Comment (jxsvoboda) [inline uspace/app/layout/layout.c:61]: Why don't you use input_open() to get an input_t object and use that to talk to the input server?

Comment (jxsvoboda) [reply uspace/app/layout/layout.c:61]: Also, print_help() is not valid syntax in ANSI/ISO C.

File: uspace/lib/c/generic/io/input.c
Comment (jxsvoboda) [inline uspace/lib/c/generic/io/input.c:213]: Why does this not have input_t * as the first argument?

File: uspace/lib/c/generic/io/input.c
Comment (jxsvoboda) [inline uspace/lib/c/generic/io/input.c:250]: Why does this not have input_t * as the first argument?

File: uspace/lib/c/include/ipc/input.h
Comment (jxsvoboda) [inline uspace/lib/c/include/ipc/input.h:42]: For symmetry with GET_LAYOUT this should be named SET_LAYOUT.

File: uspace/srv/hid/input/input.c
Comment (jxsvoboda) [inline uspace/srv/hid/input/input.c:140]: Typically the suffix used for these functions would be _srv().

File: uspace/srv/hid/input/input.c
Comment (jxsvoboda) [inline uspace/srv/hid/input/input.c:162]: This is wrong. You can't reply to *call with EOK first. You first need to call async_data_read_receive() and if it fails, you need to reply with error to call.

File: .gitignore
Comment (jxsvoboda) [inline .gitignore:8]: Why did you remove the trailing newline?

File: uspace/app/layout/layout.c
Comment (matthieuriolo) [reply uspace/app/layout/layout.c:61]: > Why don't you use input_open() to get an input_t object and use that to talk to the input server?

I remembering looking into that function and coming to the conclusion that this is an overkill (considering the refactoring it needs). The function is needed for a long time session while I need a single call.


> print_help() is not valid syntax in ANSI/ISO C

I'm sorry I cannot follow. What's wrong with the naming?

File: .gitignore
Comment (matthieuriolo) [reply .gitignore:8]: that's probably an accident from switching from the old make build to meson. My original code was written for make and I had to exclude the layout folder

File: uspace/app/layout/layout.c
Comment (le-jzr) [reply uspace/app/layout/layout.c:61]: > > print_help() is not valid syntax in ANSI/ISO C
> 
> I'm sorry I cannot follow. What's wrong with the naming?

I think what he means is that a function with no arguments is declared as `print_help(void)`. Empty parentheses mean undeclared list of arguments (or declared elsewhere), and is not accepted in modern standard C, although GNU still accepts the syntax as an extension.

**Comment by matthieuriolo** [comment]
@jxsvoboda I made the necessary changes except for
1. "Why does this not have input_t * as the first argument?"
2. "I think rather the TODO should say we should query the server for available maps"

**Comment by matthieuriolo** [comment]
> My suggestion is that, as future work, the code to look up and connect the input service should be moved to input_open() 

Ah right, so the issue is the code duplication of creating an exchange session for the input service? Because when you look at the structure of input_t and what input_open() does, they serve a completely different purpose - it's a wrapper, not just a connection. This would add complexity (passing useless NULL values to input_open()) for lil reduction of code duplication. Not sure if it is worth

**Comment by jxsvoboda** [comment]
It's not just a case of code duplication, but also of proper abstraction/encapsulation. Each service (connection to a service) should be represented by a class (translated into C, i.e. structure + functions/methods). This means the users of the API need not include <async.h> and need not know there is any IPC going on, they should be (almost) completely shielded from this. I know might not be  the case with all current code, but that's life: best practices improve and it takes time for the code to catch up. Nevertheless, when writing new code, we should observe current best practices. Such a class would typically have an interface something like:
  * method/function myservice_open(const char *svcname, myservice_t **)
  * errno_t myservice_dowhatever(myservice_t *, ...)
  * void myservice_close(myservice_t *)
I agree that these are two completely different use cases. We could make these into two separate interfaces. I.e. they could be abstracted by a different class and also separated on sever side - into two separate services.

**Review by le-jzr — APPROVED** [review]
I apologize if this was waiting for my review. It's been a crazy year.
Looks good. Just needs to resolve the new conflicts.

**Comment by matthieuriolo** [comment]
@le-jzr I rebased the branch and fixed the conflicts

**Comment by matthieuriolo** [comment]
@le-jzr @jermar @jxsvoboda guys, it's getting ridiculous. I'm tying me best to have patient and to be understanding of your situations. But it's now 2 Years... You need to find someone who takes care of the opensource part of Helenos. Obviously you are not fit for this task...

**Comment by jxsvoboda** [comment]
@matthieuriolo was that a necessary comment? A simple reminder would have sufficed. Also, if you want to get attention, a message to the mailing list is a good idea. Also a general note on contributing: If you have an idea that you'd like to implement, the *first* step that I would suggest is to discuss it on the ML so that we can first flesh out the design. It's better than shooting a pull request and than having people to deal with a change that might have fundamental design issues.

Sorry for not tending to the PRs, but one tends to avoid uncomfortable stuff (which harks back to the above).

Now back to the matter. I was almost decided to merge this and clean up the remaining issues, but when I tested it, I found out that when you change the layout using keyboard shortcuts, output of 'layout get' does not change. Then I realized the problem. Keyboard layout is currently a property of a keyboard device. You introduced layout_active, a global variable. When layout is set using 'layout' command, layouts of all keyboard devices are changed. But layouts of individual keyboard devices can be changed using keyboard shortcuts and this is not reflected in output of 'layout get'.

I consider the ability to have different layout for different keyboard device a very useful feature. Imagine having two keyboards plugged in, each with different key labels and the matching layout set in the OS.

So if the command-line layout switching is to work correctly, it should be able to list and set the layout of each individual keyboard device.

**Pull request closed without merging**
